### PR TITLE
cpu/nrf52: add support for the whole family, fix nrf52811

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The addition of nrf52811 was regrettably untested and it turns out it didn't even compile.
To generalize the fix and avoid the trap of adding CPU-specific `ifdefs`, I added support for the remaining family members (nrf52805/10/20/33) too.

I also noticed the interrupt vector tables for the existing nrf52832 and nrf52840 were wrong.

### Testing procedure

Existing nrf52 boards should still work. Especially SPI & I2C (also both at the same time),

I added a board based on the [E104-BT5011A](http://www.ebyte.com/en/product-view-news.aspx?id=1043) (nrf52810) module.
Such boards can be [bought](https://de.aliexpress.com/item/4000867497654.html) for cheap on AliExpress and similar sites. 

I can add dummy boards for the new family members - according to the documentation they can be simulated on `nrf52832dk`/`nrf52840dk`.

For now, let's see if Murdock is happy with the changes for the existing boards. 

@Citrullin, @vvzvlad do you have a nrf52811 board?

### Issues/PRs references

follow-up to #14008

### CODE REVIEW

**Comment by Citrullin** [comment]
Yes, I have. As I remember, I tested #14008 on my board. Only a simple hello world though. I will take a look into it again. Thanks for pinging. There are some links in #14008 for a nrf52811 board.

**Comment by benpicco** [comment]
I can't resist a cheap board and ordered one of [these](https://de.aliexpress.com/item/4000867497654.html) now.

**Comment by Citrullin** [comment]
> I can't resist a cheap board and ordered one of [these](https://de.aliexpress.com/item/4000867497654.html) now.
> Having the `nrf52811mdk` upstream would be beneficial for CI still.

Ohh, didn't know there is a nrf52810 already. Thanks for the link. I will also get one ^^

**Comment by benpicco** [comment]
I read about the release of the [nRF52805](https://www.nordicsemi.com/Products/Low-power-short-range-wireless/nRF52805) which lead me to wonder how much more of them are there.

**Comment by benpicco** [comment]
My board already arrived today. I haven't tried out SPI & I2C yet (and whether they work in parallel) but the rest already works. (UART, LEDs, Buttons & BLE)
Documentation for the module is a bit sparse, so I had to resort to trial & error to figure out the right pins.

**Comment by benpicco** [comment]
I can confirm that I2C and SPI do indeed work both at the same time:

```
2020-08-10 23:43:59,465 #  init 0 0 2
2020-08-10 23:43:59,472 # SPI_DEV(0) initialized: mode: 0, clk: 2, cs_port: 0, cs_pin: 0
2020-08-10 23:44:02,529 #  send test
2020-08-10 23:44:02,530 # Sent bytes
2020-08-10 23:44:02,532 #    0    1    2    3 
2020-08-10 23:44:02,534 #   0x74 0x65 0x73 0x74
2020-08-10 23:44:02,537 #     t    e    s    t 
2020-08-10 23:44:02,538 # 
2020-08-10 23:44:02,538 # Received bytes
2020-08-10 23:44:02,539 #    0    1    2    3 
2020-08-10 23:44:02,541 #   0x74 0x65 0x73 0x74
2020-08-10 23:44:02,543 #     t    e    s    t 
2020-08-10 23:44:02,544 # 
2020-08-10 23:44:07,957 #  i2c_scan 0
2020-08-10 23:44:07,958 # Scanning I2C device 0...
2020-08-10 23:44:07,960 # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2020-08-10 23:44:07,961 #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2020-08-10 23:44:07,961 # 0x00 R R R R R R R R R R R R R R - -
2020-08-10 23:44:07,966 # 0x10 - - - - - - - - - - - - - - - -
2020-08-10 23:44:07,971 # 0x20 - - - - - - - - - - - - - - - -
2020-08-10 23:44:07,977 # 0x30 - - - - - - - - - - - - - - - -
2020-08-10 23:44:07,982 # 0x40 - - - - - - - - - - - - - - - -
2020-08-10 23:44:07,987 # 0x50 - - - X - - - - - - - - - - - -
2020-08-10 23:44:07,993 # 0x60 - - - - - - - - X - - - - - - -
2020-08-10 23:44:07,997 # 0x70 - - - - - - - - R R R R R R R R
```

(Sorry, didn't mean to rebase. I rebased because I hoped I now could flash the board using ST-Link & OpenOCD, but looks like OpenOCD doesn't like the chip yet. Works with JLink & JLinkExe. Still pushed…)

File: boards/e104-bt5010a-tb/Kconfig [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2020 HAW Hamburg
```
Comment (fjmolinas) [inline boards/e104-bt5010a-tb/Kconfig:1]: Is this the right copyright?

File: boards/e104-bt5010a-tb/include/periph_conf.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * @name    Clock configuration
+ *
+ * @note    The radio will not work with the internal RC oscillator!
+ *
+ * @{
+ */
+#define CLOCK_HFCLK         (0U)            /* set to  0: internal RC oscillator
+                                             *        32: 32MHz crystal */
+#define CLOCK_LFCLK         (0)             /* set to  0: internal RC oscillator
```
Comment (fjmolinas) [inline boards/e104-bt5010a-tb/include/periph_conf.h:40]: Can you add `ifndef` guards here?

**Review by fjmolinas — COMMENTED** [review]
Some comments below, also why do you think ea2638dcace09bbf5154f2d84129253a7155acaf is a more maintainable approach? It seems so far the approach for other families has been to have a unified file right? (I don't really mind, just asking for the reasoning.)

**Comment by fjmolinas** [comment]
@benpicco can you provide some more test output for the new `BOARD`?

**Comment by benpicco** [comment]
> why do you think ea2638d is a more maintainable approach?

I noticed the file would turn into a thick ifdef soup.
This was worsened by the fact that some MCUs share IRQs for *some* SPI and I2C buses (e.g. SPI2 and I2C1 share the same IRQ, but another MCU doesn't have SPI2 so the IRQ is for I2C1 only)
We would just use the same name everywhere of course, but that would require touching even more code AFAIR.

#14595 would probably be the best approach here too.

**Comment by fjmolinas** [comment]
> #14595 would probably be the best approach here too.

Ok lets go with the current approach for now.

**Comment by fjmolinas** [comment]
> I can confirm that I2C and SPI do indeed work both at the same time (I added i2c_scan to tests/periph_spi):

Is this on the new board? So I test for regressions for the other cpus

**Comment by benpicco** [comment]
Yea that's on the 'smaller' family members where certain features/peripherals are cut.

**Comment by fjmolinas** [comment]
Tested for regression by initiating the `dw1000` (spi) and the `lisdh12` on board of the `dwm1001` (`nrf52832`)

```
2020-08-25 13:43:39,977 #  ifconfig
2020-08-25 13:43:39,987 # Iface  3  HWaddr: 12:6C  Channel: 0  Frequency: 1024Hz  Page: 0  NID: 0x0  RSSI: 4  PHY: DISABLED  Link: down
2020-08-25 13:43:39,991 #           Long HWaddr: 1E:88:5A:5D:6A:E6:9C:7D
2020-08-25 13:43:39,997 #            TX-Power: 8192dBm  State: IDLE  max. Retrans.: 0
2020-08-25 13:43:40,001 #           L2-PDU:102  Source address length: 2
2020-08-25 13:43:40,002 #
> saul
2020-08-25 13:43:44,315 #  saul
2020-08-25 13:43:44,315 #
2020-08-25 13:43:44,318 # Dev: NRF_TEMP Type: SENSE_TEMP
2020-08-25 13:43:44,320 # Data:              31 °C
2020-08-25 13:43:44,322 # [i2c] acquired dev 0
2020-08-25 13:43:44,327 # [i2c] read_regs: 1 byte(s) from reg 0x27 at addr 0x19
2020-08-25 13:43:44,330 # [i2c] waiting for STOPPED or ERROR event
2020-08-25 13:43:44,333 # [i2c] finish: stop event occurred
2020-08-25 13:43:44,338 # [i2c] read_regs: 6 byte(s) from reg 0xa8 at addr 0x19
2020-08-25 13:43:44,341 # [i2c] waiting for STOPPED or ERROR event
2020-08-25 13:43:44,344 # [i2c] finish: stop event occurred
2020-08-25 13:43:44,346 # [i2c] released dev 0
2020-08-25 13:43:44,346 #
2020-08-25 13:43:44,349 # Dev: lis2dh12 Type: SENSE_ACCEL
2020-08-25 13:43:44,351 # Data: [0]         -42 mg
2020-08-25 13:43:44,353 #       [1]          35 mg
2020-08-25 13:43:44,355 #       [2]         902 mg
```

**Comment by benpicco** [comment]
> Tested for regression by initiating the `dw1000` (spi) and the `lisdh12` on board of the `dwm1001` (`nrf52832`)

There is a netdev driver for that :eyes:

**Comment by fjmolinas** [comment]
> There is a netdev driver for that eyes

I'm working on it.

**Comment by fjmolinas** [comment]
Tested that there is not regression on `nrf52` by running `examples/filesystem` + `i2c_scan` on `nrf52840_dk` (that uses `mtd_spi_nor`).

```
2020-08-25 14:07:24,921 # main(): This is RIOT! (Version: 2020.10-devel-987-g78140e-wip/dw1000)
2020-08-25 14:07:24,924 # constfs mounted successfully
i2c_scan 0
2020-08-25 14:07:26,807 #  i2c_scan 0
2020-08-25 14:07:26,810 # Scanning I2C device 0...
2020-08-25 14:07:26,816 # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2020-08-25 14:07:26,819 #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2020-08-25 14:07:26,823 # 0x00 R R R R R R R R R R R R R R - -
2020-08-25 14:07:26,828 # 0x10 - - - - - - - - - - - - - - - -
2020-08-25 14:07:26,833 # 0x20 - - - - - - - - - - - - - - - -
2020-08-25 14:07:26,839 # 0x30 - - - - - - - - - - - - - - - -
2020-08-25 14:07:26,844 # 0x40 - - - - - - - - - - - - - - - -
2020-08-25 14:07:26,849 # 0x50 - - - - - - - - - - - - - - - -
2020-08-25 14:07:26,854 # 0x60 - - - - - - - - - - - - - - - -
2020-08-25 14:07:26,859 # 0x70 - - - - - - - X R R R R R R R R
```

**Comment by fjmolinas** [comment]
@benpicco once you provide some more test output for the new board and address https://github.com/RIOT-OS/RIOT/pull/14372#discussion_r476328781 and https://github.com/RIOT-OS/RIOT/pull/14372#discussion_r476327860, this one is good to go.

File: boards/e104-bt5010a-tb/include/periph_conf.h (hunk 2 above)
Comment (benpicco) [reply boards/e104-bt5010a-tb/include/periph_conf.h:40]: Those are physical properties of the board (module).
 There is neither a 32kHz oscillator nor a 32 MHz one present on the module, so no point in overwriting this.

**Comment by benpicco** [comment]
What test output would you like to see?

File: boards/e104-bt5010a-tb/include/periph_conf.h (hunk 2 above)
Comment (fjmolinas) [reply boards/e104-bt5010a-tb/include/periph_conf.h:40]: Ah ok, can you fix the comment then? The comment on the side made me think they where configurable.

Comment (benpicco) [reply boards/e104-bt5010a-tb/include/periph_conf.h:40]: Uh yea

> The radio will not work with the internal RC oscillator!

is obviously wrong.

**Comment by fjmolinas** [comment]
> What test output would you like to see?

Ideally `compile_and_test_for_board`.

**Comment by benpicco** [comment]
Flashing is a bit unreliable with my setup :weary: 

I just managed to flash `examples/nimble_scanner` by holding the board in my hands… 

```
2020-08-25 14:44:51,885 #  main(): This is RIOT! (Version: 2020.10-devel-665-g6540a-cpu/nrf52-full)
2020-08-25 14:44:51,886 # NimBLE Scanner Example Application
2020-08-25 14:44:51,887 # Type `scan help` for more information
2020-08-25 14:45:05,271 #  scan
2020-08-25 14:45:06,276 # Scanning for 1000ms now ... done
2020-08-25 14:45:06,276 # 
2020-08-25 14:45:06,276 # Results:
2020-08-25 14:45:06,287 # [ 0] 35:0e:13:8d:d0:7b (RANDOM) [NONCONN_IND] "undefined", adv_msg_cnt: 4, adv_int: 205135us, last_rssi: -49
```

**Comment by fjmolinas** [comment]
> Flashing is a bit unreliable with my setup weary

Ok I guess, we can count `spi`, `i2c`, `timer` and `uart`, tested with this. Although in general I prefer having the complete suit of tests ran on new `BOARD`s. Lets see if murdock is happy.

**Comment by benpicco** [comment]
Murdock looks reasonably happy :)

**Comment by fjmolinas** [comment]
@benpicco please squash, I'll run all tests on `nrf52` boards just in case.

**Review by fjmolinas — APPROVED** [review]
ACK, ran all tests on an nrf52840 [master](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/165/artifact/) [pr](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/164/artifact/) and there where no new test failures. Although there where some unexpected failures that where also present in master that I'll have to investigate.

**Comment by benpicco** [comment]
Thank you for the review!

**Comment by Citrullin** [comment]
@benpicco We talked at the summit. I think you are right. Just remove my name in the files when you update them at some point :)

**Pull request merged**